docs/iid-1.md

```diff
@@ -0,0 +1,297 @@
+
+##### **BoringCrypto**
+BoringCrypto (BoringSSL based crypto) maintained by Google is an open-source, general-purpose cryptographic library that provides FIPS 140–2 approved cryptographic algorithms to serve BoringSSL and other user-space applications BoringSSL is Google’s forked version of OpenSSL cryptographic library.
+
+##### Limitations of BoringCrypto
+
+- The build must be GOOS=linux, GOARCH=amd64.
+- The build must have cgo enabled.
+- The android build tag must not be specified.
+- The cmd_go_bootstrap build tag must not be specified.
+- The version string reported by runtime.Version does not indicate that BoringCrypto was actually used for the build. For example, linux/386 and non-cgo linux/amd64 binaries will report a version of go1.8.3b2 but not be using BoringCrypto.
+
+Please note, currently there is no support for arm64 builds for BoringCrypto. Refer [here](https://github.com/golang/go/issues/39760) for the same.
+
+### What does BoringCrypto actually do?
+
+Golang has a crypto standard library, Go `crypto` package which supports various crypto requirements. (TLS implementation HTTPS servers/clients, libraries for signatures to verify hashes, encrypt/decrypt messages)
+
+TLS protocol supports the use of numerous cryptographic protocols. The knowledge of which protocol (or cipher suites) to use is important. See [cloudflare](https://blog.cloudflare.com/exposing-go-on-the-internet/) post on this. [Mozilla](https://wiki.mozilla.org/Security/Server_Side_TLS) provides a document is to help operational teams with the configuration of TLS, which is widely used.
+
+Since the native go crypto is not FIPS compliant because it allows the usage of multiple algorithms, protocols, ciper suites and versions (which are vulnerable), BoringCrypto restrics this and provides FIPS 140–2 approved cryptographic algorithms to serve BoringSSL and other user-space applications.
+
+BoringCrypto internally uses cgo to execute calls to BoringSSL to implement various crypto primitives: https://github.com/golang/go/blob/dev.boringcrypto/README.boringcrypto.md.
+
+The BoringCrypto library provides a C-language API for use by other processes that require cryptographic functionality. All operations of the module occur via calls from host applications and their respective internal daemons/processes. The cryptographic boundary of the BoringCrypto module is a single object file named bcm.o which is statically linked to BoringSSL. The library performs no communications other than with the calling application (the process that invokes the library services) and the host operating system.
+
+
+### Implementation
+
+Implementation proposal for **immudb** FIPS binary is by creating a new `Dockerfile.fips` which builds from the boringcrypto golang image. This helps in the following ways:
```

Comment:
I think we may also need to adjust the Makefile to produce fips-compliant binaries without using the dockerfile.

There's `make dist/all` Makefile target that should build immudb binaries for all platforms/os-es - I believe we should add fips-compliant builds there too. By doing so we will automatically notarize those with current GH actions and it will be easier to smoke-test them and upload when doing the release.

Such pre-built immudb binaries would require the target system to contain boringssl libraries thus some additional documentation will be needed on how to run them.

Reply:
ack, I'll do that